src/constraint/SequentialSampleable.cpp

```diff
@@ -0,0 +1,174 @@
+#include "aikido/constraint/SequentialSampleable.hpp"
+
+#include <dart/common/Console.hpp>
+#include <dart/common/Memory.hpp>
+
+namespace aikido {
+namespace constraint {
+
+//==============================================================================
+class SequentialSampleGenerator : public SampleGenerator
+{
+public:
+  /// Sample Generator for Sequential Sampleable
+  /// \param[in] stateSpace Statespace the associated constraint operates in
+  /// \param[in] generators Sequence of generators associated with corresponding
+  /// sequence of sampleables
+  SequentialSampleGenerator(
+      statespace::StateSpacePtr stateSpace,
+      std::vector<std::unique_ptr<SampleGenerator>> generators);
+
+  // Documentation inherited
+  statespace::StateSpacePtr getStateSpace() const override;
+
+  // Documentation inherited
+  bool sample(statespace::StateSpace::State* state) override;
+
+  // Documentation inherited
+  int getNumSamples() const override;
+
+  // Documentation inherited
+  bool canSample() const override;
+
+private:
+  /// StateSpace the associated sampleable operates in.
+  statespace::StateSpacePtr mStateSpace;
+
+  /// Sequence of generators associated with corresponding sequence of
+  /// sampleables.
+  std::vector<std::unique_ptr<SampleGenerator>> mGenerators;
+
+  /// Index of the active sampleable/generator.
+  std::size_t mIndex;
+};
+
+//==============================================================================
+SequentialSampleGenerator::SequentialSampleGenerator(
+    statespace::StateSpacePtr stateSpace,
+    std::vector<std::unique_ptr<SampleGenerator>> generators)
+  : mStateSpace(std::move(stateSpace))
+  , mGenerators(std::move(generators))
+  , mIndex(0)
+{
+  for (std::size_t i = 0; i < mGenerators.size(); ++i)
+  {
+    if (!mGenerators[i])
+    {
+      std::stringstream msg;
+      msg << "Generator " << i << " is nullptr.";
+      throw std::invalid_argument(msg.str());
+    }
+  }
+
+#ifndef NDEBUG
+  for (const auto& generator : mGenerators)
+    assert(generator->getStateSpace() == mStateSpace);
+#endif
+
+  for (std::size_t i = 0u; i < mGenerators.size() - 1; ++i)
+  {
+    if (mGenerators[i]->getNumSamples() == NO_LIMIT)
+      dtwarn << "Sampleable " << i << " is infinite. "
+             << "Remaining " << mGenerators.size() - i - 1
+             << " sampleable(s) will potentially be ignored.";
```

Comment:
This warning seems very informative! It would be slightly better:
* to `break` once print a warning
* not to warn if `i` is equal to `mGenerators.size() - 1`; otherwise, `mGenerators.size() - i - 1` would be the greatest number of `std::size_t`.

Reply:
Agree with the first point, but I think the second point should never happen since the for loop condition is `i < mGenerators.size() - 1`?

Reply:
Oh, I missed that. Please ignore the second point.